### PR TITLE
`gr-icon` directive.

### PR DESCRIPTION
**Pull request by akash1810**
Typing `<i class="material-icons">icon</i>` is long...

TODO
- [ ] icons everywhere - looking at you ▸ and ▾

### CODE REVIEW

File: kahuna/public/js/preview/image.html [hunk 1]
```diff
@@ -61,7 +61,7 @@
                   title="this image has crops"
                   ng:if="image | hasExportsOfType:'crop'">
 
-                <i class="material-icons material-icons--small">crop</i>
+                <gr-icon small="true">crop</gr-icon>
```
Comment (akash1810) [inline kahuna/public/js/preview/image.html]: couldn't work out a way to use html5 attributes, i.e. `<gr-icon small>crop</gr-icon>`

File: kahuna/public/js/components/gu-date-range/gu-date-range.html [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <div class="gu-date-range" ng:init="ctrl.showOverlay = false">
     <div class="gu-date-range__display" ng:click="ctrl.showOverlay = !ctrl.showOverlay">
         <div class="gu-date-range__display__icon gu-date-range__display__icon__calendar">
-            &#128197;
+            <gr-icon>event</gr-icon>
```
Comment (akash1810) [inline kahuna/public/js/components/gu-date-range/gu-date-range.html]: bye bye emoji date icon!

File: kahuna/public/js/components/gu-date-range/gu-date-range.html [hunk 3]
```diff
@@ -1,9 +1,6 @@
 <div class="gu-date-range" ng:init="ctrl.showOverlay = false">
     <div class="gu-date-range__display" ng:click="ctrl.showOverlay = !ctrl.showOverlay">
-        <div class="gu-date-range__display__icon gu-date-range__display__icon__calendar">
-            &#128197;
-        </div>
-
+        <gr-icon>event</gr-icon>
```
Comment (akash1810) [inline kahuna/public/js/components/gu-date-range/gu-date-range.html]: bye bye calendar emoji icon

File: kahuna/public/js/components/gr-icon/gr-icon.css [hunk 4]
```diff
@@ -0,0 +1,27 @@
+.gr-icon {
```
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-icon/gr-icon.css]: Could we not just style the `gr-icon` element.... @theefer was using that as a pattern elsewhere and it's quite nice.

**Comment by jamesgorrie** [comment]
Have you included this in all the component where it's used?

**Comment by akash1810** [comment]
Yep - cheated and [put it in the root](https://github.com/guardian/grid/pull/941/files#diff-a727a260b31fe8cd84ba0780a0ee0370R62) so its only required once.

**Comment by jamesgorrie** [comment]
Yea - I was wondering if that's the right thing to do?
That way if a component is removed from the root scope, it'll break - do you think it adds to much gumpf to add to each directive?

File: kahuna/public/js/components/gr-icon/gr-icon.css (hunk 4 above)
Comment (akash1810) [inline kahuna/public/js/components/gr-icon/gr-icon.css]: Could do, would mean we'd also have to have

```css
gr-icon i {
    font-style: normal;
}
```

To override the default browser styling of `<i>`.

File: kahuna/public/js/preview/image.html (hunk 1 above)
Comment (jamesgorrie) [inline kahuna/public/js/preview/image.html]: You might want to call the attr gr:small too?

Comment (jamesgorrie) [inline kahuna/public/js/preview/image.html]: You could remove the scope and have the link:
```JavaScript
{
  link: function(scope, elem, attrs) { scope.small = angular.exists(attrs.grSmall) }
}
```

**Comment by akash1810** [comment]
> do you think it adds to much gumpf to add to each directive?

Yeah. You'd end up doing `import 'components/gr-icon/gr-icon';` and `angular.module('myModule', ['grIcon']);` everywhere you want to icon, which is more cumbersome than just `<i class="material-icons">icon</i>` in a template.

File: kahuna/public/js/components/gr-icon/gr-icon.css (hunk 4 above)
Comment (akash1810) [inline kahuna/public/js/components/gr-icon/gr-icon.css]: Done. Targeting `gr-icon .gr-icon` to avoid `gr-icon i`.

**Comment by jamesgorrie** [comment]
:+1: it'd be interesting to chat about BEM and the structure of our CSS, as in a componentalised world I'm not sure it makes the most sense.

**Comment by kenoir** [comment]
:+1: nice

**Comment by akash1810** [comment]
> it'd be interesting to chat about BEM and the structure of our CSS, as in a componentalised world I'm not sure it makes the most sense.

Yeah. I guess it depends on the size of the components; though you could argue that large components should be split up into smaller ones...

**Pull request merged**